xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/AudioDecoder.h

```diff
@@ -0,0 +1,242 @@
+    /// @ingroup cpp_kodi_addon_audiodecoder
+    /// @brief Get number of tracks in a file
+    ///
+    /// @param[in] file                 File to read tag for
+    /// @return                         Number of tracks in file
+    ///
+    virtual int TrackCount(std::string file) { return 1; }
+    //--------------------------------------------------------------------------
+
+  private:
+    void SetAddonStruct(KODI_HANDLE instance)
+    {
+      if (instance == nullptr)
+        throw std::logic_error("kodi::addon::CInstanceAudioDecoder: Creation with empty addon structure not allowed, table must be given from Kodi!");
+
+      m_instanceData = static_cast<AddonInstance_AudioDecoder*>(instance);
+
+      m_instanceData->toAddon.addonInstance = this;
+      m_instanceData->toAddon.init = ADDON_Init;
+      m_instanceData->toAddon.read_pcm = ADDON_ReadPCM;
+      m_instanceData->toAddon.seek = ADDON_Seek;
+      m_instanceData->toAddon.read_tag = ADDON_ReadTag;
+      m_instanceData->toAddon.track_count = ADDON_TrackCount;
+    }
+
+    inline static bool ADDON_Init(const AddonInstance_AudioDecoder* instance, const char* file, unsigned int filecache,
+                                  int* channels, int* samplerate,
+                                  int* bitspersample, int64_t* totaltime,
+                                  int* bitrate, AEDataFormat* format,
+                                  const AEChannel** info)
+    {
+      instance->toAddon.addonInstance->m_channelList.clear();
+      bool ret = instance->toAddon.addonInstance->Init(file, filecache, *channels,
+                                                           *samplerate, *bitspersample,
+                                                           *totaltime, *bitrate, *format,
+                                                           instance->toAddon.addonInstance->m_channelList);
+      if (!instance->toAddon.addonInstance->m_channelList.empty())
+      {
+        if (instance->toAddon.addonInstance->m_channelList.back() != AE_CH_NULL)
+          instance->toAddon.addonInstance->m_channelList.push_back(AE_CH_NULL);
+        *info = instance->toAddon.addonInstance->m_channelList.data();
+      }
+      else
+        *info = nullptr;
+      return ret;
+    }
+
+    inline static int ADDON_ReadPCM(const AddonInstance_AudioDecoder* instance, uint8_t* buffer, int size, int* actualsize)
+    {
+      return instance->toAddon.addonInstance->ReadPCM(buffer, size, *actualsize);
+    }
+
+    inline static int64_t ADDON_Seek(const AddonInstance_AudioDecoder* instance, int64_t time)
+    {
+      return instance->toAddon.addonInstance->Seek(time);
+    }
+
+    inline static bool ADDON_ReadTag(const AddonInstance_AudioDecoder* instance, const char* file, char* title, char* artist, int* length)
+    {
+      std::string intTitle;
+      std::string intArtist;
+      memset(title, 0, ADDON_STANDARD_STRING_LENGTH_SMALL);
+      memset(artist, 0, ADDON_STANDARD_STRING_LENGTH_SMALL);
```

Comment:
Think this only work on a static structure where the size can easily identified.